ADR/0033-upstreaming-integration-service.md

```diff
@@ -0,0 +1,270 @@
+a [selector] which the user can use to reference the `LABELNAME` label on [TestSubjects]. It should
+be possible for the user to specify a selector that references any attribute they wish. The user
+should conclude naturally that `LABELNAME` is a good candidate for usage in their selectors. UI and
+docs can guide this.
+
+- A new [TestSubject] without a `LABELNAME` label should be mutated by webhook to gain a value of
+  "default".
+- Attempts to remove the `LABELNAME` label from a [TestSubject] should be rejected (or mutated to
+  "default"?)
+- Attempts to edit the `LABELNAME` label on a [TestSubject] should be permitted.
+
+### The Global Candidate List becomes the control TestSubject, the last TestSubject to pass testing
+
+Today, the "global candidate list" is a term that we use to refer to the list of all the built-image
+pullspecs attached to every [Component] of an [Application]. Replace that implementation
+with an identifying *label* on a [TestSubject] in the namespace: `CONTROLLABEL`. Call the [TestSubject] that
+carries this `CONTROLLABEL` label the **control** TestSubject for the `LABELNAME` label.
+
+[integration-service] should grow a mutating webhook that *removes* the `CONTROLLABEL` label from
+the last control [TestSubject] for the `LABELNAME` label, when the `CONTROLLABEL` label is applied
+to a new [TestSubject] bearing the `LABELNAME` label.
+
+The **control** TestSubject is used as the *basis* for new TestSubject construction. See
+"TestSubject Construction as optional behavior", below.
+
+[integration-service] should promote a [TestSubject] under test to become the new **control**
+TestSubject after it passes all of the required IntegrationTestScenarios with the matching
+`LABELNAME` label.
+
+### TestSubject Construction as optional behavior
+
+**Test Initiation** has a clear control plane: the [IntegrationTestScenario] combined with the
+[TestSubject]. Meaning, when the user creates an [IntegrationTestScenario] and a [TestSubject], and
+the [selector] of the [IntegrationTestScenario] matches the attributes of [TestSubject], then together
+they represent intent for the subject to be tested. The user uses these resources to control what
+the system does.
+
+Today, control of the **Snapshot Construction** process is mixed up in the [Application] and [Component]
+model, and is partly just implied. We *always* construct [Snapshots].
+
+**TestSubject Construction** (called Snapshot Construction in our system today) should have a clear
+control plane resource: the [TestSubjectConstructor]. A hypothetical schema for the resource
+follows:
+
+```yaml
+version: appstudio.redhat.com/v1alpha1
+kind: TestSubjectConstructor
+metadata:
+  name: my-application-test-subject-constructor
+spec:
+  selector:
+    fields:
+      match:
+        kind: PipelineRun
+        version: tekton.dev/v1
+    labels:
+      match:
+        "appstudio.openshift.io/application": my-application
+        "pipelinesascode.tekton.dev/state": completed
```

Comment:
I know that this is just illustrative, but these fields seem like they might not be narrowly defined enough. Would these also match on the IntegrationTestScenarios triggered as well as it seems like we would want to associate those with an application as well.

Reply:
Might an IntegrationTestScenario have a label associating the related application? If the IntegrationTestScenarios are additionally triggered by PAC (or if there are non-build PAC pipelines), then we could enter into a situation where the TestSubjectConstructor reacts to more PipelineRuns than just ones that produce artifacts.